### ISSUE TITLE
Testing Multi-Instance with Collection

### ISSUE THREAD

**Issue by bennobuilder** [camunda-community-hub/bpmn-driven-testing#192]
I’m currently working on a test case using BPMN-Driven-Testing and encountered an issue while trying to test a multi-instance subprocess configured with a collection (`camunda:collection`, see image). 

In my test, I expect the `verifyLoopCount` to reflect the size of the collection (in this case, `2` for a list of two emails), but it always returns `0`. The variables are being consumed (-> to verify ProcessInstance contains variable `emails` but empty). 

Is there support for testing `camunda:collection` in this context (couldn't find any test) or is only loop cardinality supported. An example or guidance would be appreciated.

Thanks :)

![image](https://github.com/user-attachments/assets/635e0cf8-f02e-494e-b3fb-8ef1877bf106)

```java
@SpringBootTest(classes = {Application.class, BpmndtConfiguration.class})
class EmailBounceHandlingProcessTest {

  @RegisterExtension
  TC_EmailBounceHandlingProcess_TC1 tc1 = new TC_EmailBounceHandlingProcess_TC1();

  @Test
  void testTC1() {
    var emailsVariable = List.of("email1.eml", "email2.eml");

    tc1.handleCallProcessEveryMessageSubprocess().verifyParallel();
    // TODO: Can we test "camunda:collection" with BPMN-Driven-Testing
    // tc1.handleCallProcessEveryMessageSubprocess().verifyLoopCount(emailsVariable.size());
//    tc1.handleCallProcessEveryMessageSubprocess().handle(0).verify((pi, callActivity) -> {
//      pi.variables().containsKey("email");
//    });
//    tc1.handleCallProcessEveryMessageSubprocess().handle(1).verify((pi, callActivity) -> {
//      pi.variables().containsKey("email");
//    });

    tc1.createExecutor()
       .withBean("collectAllEmails", new DelegateMock((new HashMap<String, Object>() {{
         put("emails", emailsVariable);
       }})))
       .verify(pi -> {
         pi.variables()
           .containsKey("emails"); // empty list
         pi.isEnded();
       })
       .execute();
  }
}
```

**Comment by gclaussn** [comment]
hi @bennobuilder,

on branch issue/192 (see [commit](https://github.com/camunda-community-hub/bpmn-driven-testing/commit/054a816d4d20101f512f267e22e1aeb9aac1ed4f)) I added an example process with a multi instance call activity that uses a collection expression `${elements}` and element variable `element`. A test starts a test case with variable elements=["a", "b", "c"] and verifies the loop count as well as the variables at the call activity within the multi instance.

it is a bit tricky to test the local element variable, used within the multi instance.

in the `verify` method of the call activity handler of the multi instance a `ProcessInstanceAssert` (here named `piAssert`) is provided. but it can only be used to assert variables that are on process instance level, e.g. `elements`. to verify local variables, like the element variable, the local variables of the related exectution must be retrieved. to do this, the execution ID is required. therefore the historic activity instance of the multi instance body must be found.
```java
    tc.handleCallActivity().handle(0).verify((piAssert, callActivityDefinition) -> {
        piAssert.variables().containsKey("elements"); // assertion of process scoped variables only (e.g. collection)

        // find process instance of test case, e.g. by business key
        var pi = tc.getProcessEngine().getRuntimeService().createProcessInstanceQuery()
            .processInstanceBusinessKey("bk")
            .singleResult();

        // find historic activity instance of call activity in multi instance scope
        var historicActivityInstance = tc.getProcessEngine().getHistoryService().createHistoricActivityInstanceQuery()
              .activityId("multiInstanceCallActivity#multiInstanceBody") // BPMN element ID + #multiInstanceBody
              .singleResult();

        // get local variables (including element variable of collection) of related execution
        var localVariables = tc.getProcessEngine().getRuntimeService().getVariablesLocal(historicActivityInstance.getExecutionId());
        assertThat(localVariables.get("element"), equalTo("a"));
    });
```

when the element variable is propagated to the subprocess via delegate variable mapping, in mapping propagation or in mappings, the variable value is much easier to assert.
```java
    tc.handleCallActivity().handle(0).verifyInput(variables -> {
        assertThat(variables.getVariable("element"), equalTo("a"));
    });
    tc.handleCallActivity().handle(1).verifyInput(variables -> {
        assertThat(variables.getVariable("element"), equalTo("b"));
    });
    tc.handleCallActivity().handle(2).verifyInput(variables -> {
        assertThat(variables.getVariable("element"), equalTo("c"));
    });
```

this example shows that a multi instance call activity with collection works fine.

regarding your test, I'm not sure what `DelegateMock` is doing. maybe the variables is not set correctly.

if you have more questions, feel free to ask.

best regards!

**Comment by bennobuilder** [comment]
Hi @gclaussn,

Thanks for the detailed explanation and example. After further investigation, I identified two issues that contributed to my failing tests.

## Discovery 1: Variable Overriding by Output

I found that defining `Outputs` in a Delegate expression can override initially set variables (e.g., by `tc1.createExecutor().withVariable(..)`). This caused the issue I initially encountered, where I expected variables to persist but found they were overwritten by an Output (without values) defined in the modeler (see `Screenshot 1`).

<details>
<summary>Screenshot 1</summary>

![image](https://github.com/user-attachments/assets/a6b859a4-1a67-4446-a292-7f90a5da3168)

</details>

## Discovery 2: Different Behavior with Spring Boot

The `bpmn-driven-testing` library behaves differently when integrated with Spring Boot compared to a plain Java setup.

### With Spring Boot

Running multiple test cases simultaneously results in exceptions like `OptimisticLockingException` and `ProcessEngineException` (see `Screenshot 2`):

```java
org.camunda.bpm.engine.ProcessEngineException: ENGINE-14026 No job found with id 'a35e4042-66c8-11ef-a37f-00059a3c7a00'

org.camunda.bpm.engine.OptimisticLockingException: ENGINE-03005 Execution of 'DELETE MessageEntity[b9969346-66c8-11ef-9cb8-00059a3c7a00]' failed. Entity was updated by another transaction concurrently.
```

<details>
<summary>Screenshot 2</summary>

![image](https://github.com/user-attachments/assets/c3576d87-52b3-45ff-a506-3a14b961f7d8)

</details>

Running only the first test case (similar to my initial test case but simplified, see `Test Case`) results in an `AssertionError` (see `Screenshot 3`):

```java
java.lang.AssertionError: Expected at least one job for activity 'Call_Participant_2'
```

<details>
<summary>Screenshot 3</summary>

![image](https://github.com/user-attachments/assets/365577d5-9f8d-4124-98d1-eca7a2cff092)

</details>

### Without Spring Boot

When I remove Spring Boot (i.e., the `SpringBootTest` annotation), the tests run as expected without exceptions (see `Screenshot 4`). I confirmed this by setting the expected loop count to 2 and anticipating an error.

<details>
<summary>Screenshot 4</summary>

![image](https://github.com/user-attachments/assets/c63515d2-0464-46c8-b4a1-648d5ef662ed)

</details>

## Test Case

These issues were observed using the following files, added to the [`advanced-spring-boot`](https://github.com/camunda-community-hub/bpmn-driven-testing/tree/master/integration-tests/advanced-spring-boot) integration tests (checked out locally on Windows 11 with Java 21):

<details>
<summary>participants.bpmn</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" xmlns:bpmndt="http://camunda.org/schema/extension/bpmn-driven-testing" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_1wdpnds" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.22.0" modeler:executionPlatform="Camunda Platform" modeler:executionPlatformVersion="7.21.0">
  <bpmn:collaboration id="Collaboration_045oqf3">
    <bpmn:participant id="Participant_1" name="Participant 1" processRef="Participant_1_PID" />
    <bpmn:participant id="Participant_3" name="Participant 3" processRef="Participant_3_PID" />
    <bpmn:participant id="Participant_2" name="Participant 2" processRef="Participant_2_PID" />
  </bpmn:collaboration>
  <bpmn:process id="Participant_1_PID" name="Participant 1 Process" isExecutable="true" camunda:versionTag="1" camunda:historyTimeToLive="30">
    <bpmn:extensionElements>
      <bpmndt:testCases>
        <bpmndt:testCase>
          <bpmndt:name>Participant1_TC1</bpmndt:name>
          <bpmndt:path>
            <bpmndt:node>StartEvent_Participant_1</bpmndt:node>
            <bpmndt:node>CollectElementsTask</bpmndt:node>
            <bpmndt:node>Call_Participant_2</bpmndt:node>
            <bpmndt:node>EndEvent_Participant_1</bpmndt:node>
          </bpmndt:path>
        </bpmndt:testCase>
      </bpmndt:testCases>
    </bpmn:extensionElements>
    <bpmn:laneSet id="LaneSet_17bjefu" />
    <bpmn:startEvent id="StartEvent_Participant_1" name="Every hour">
      <bpmn:outgoing>Flow_19m8hkc</bpmn:outgoing>
      <bpmn:timerEventDefinition id="TimerEventDefinition_1efxlu2">
        <bpmn:timeCycle xsi:type="bpmn:tFormalExpression">0 0 1 1-31 1-12 ?</bpmn:timeCycle>
      </bpmn:timerEventDefinition>
    </bpmn:startEvent>
    <bpmn:endEvent id="EndEvent_Participant_1" name="End">
      <bpmn:incoming>Flow_1hqadxe</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:callActivity id="Call_Participant_2" name="Call Participant 2" calledElement="Participant_2_PID">
      <bpmn:incoming>Flow_0q1lqbh</bpmn:incoming>
      <bpmn:outgoing>Flow_1hqadxe</bpmn:outgoing>
      <bpmn:multiInstanceLoopCharacteristics camunda:collection="${elements}" camunda:elementVariable="element" />
    </bpmn:callActivity>
    <bpmn:serviceTask id="CollectElementsTask" name="Collect Elements" camunda:delegateExpression="${collectElements}">
      <bpmn:extensionElements />
      <bpmn:incoming>Flow_19m8hkc</bpmn:incoming>
      <bpmn:outgoing>Flow_0q1lqbh</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_19m8hkc" sourceRef="StartEvent_Participant_1" targetRef="CollectElementsTask" />
    <bpmn:sequenceFlow id="Flow_1hqadxe" sourceRef="Call_Participant_2" targetRef="EndEvent_Participant_1" />
    <bpmn:sequenceFlow id="Flow_0q1lqbh" sourceRef="CollectElementsTask" targetRef="Call_Participant_2" />
  </bpmn:process>
  <bpmn:process id="Participant_3_PID" name="Participant 3 Process" isExecutable="true" camunda:versionTag="1" camunda:historyTimeToLive="30">
    <bpmn:extensionElements>
      <bpmndt:testCases>
        <bpmndt:testCase>
          <bpmndt:name>Participant3_TC1</bpmndt:name>
          <bpmndt:path>
            <bpmndt:node>StartEvent_Participant_3</bpmndt:node>
            <bpmndt:node>UserTaks_1</bpmndt:node>
            <bpmndt:node>EndEvent_Participant_3</bpmndt:node>
          </bpmndt:path>
        </bpmndt:testCase>
      </bpmndt:testCases>
    </bpmn:extensionElements>
    <bpmn:startEvent id="StartEvent_Participant_3" name="Start">
      <bpmn:outgoing>Flow_08dpap9</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:userTask id="UserTaks_1" name="UserTask">
      <bpmn:incoming>Flow_08dpap9</bpmn:incoming>
      <bpmn:outgoing>Flow_00eq203</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:endEvent id="EndEvent_Participant_3" name="End">
      <bpmn:incoming>Flow_00eq203</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_08dpap9" sourceRef="StartEvent_Participant_3" targetRef="UserTaks_1" />
    <bpmn:sequenceFlow id="Flow_00eq203" sourceRef="UserTaks_1" targetRef="EndEvent_Participant_3" />
  </bpmn:process>
  <bpmn:process id="Participant_2_PID" name="Participant 2 Process" isExecutable="true" camunda:versionTag="1" camunda:historyTimeToLive="30">
    <bpmn:extensionElements>
      <bpmndt:testCases>
        <bpmndt:testCase>
          <bpmndt:name>Participant2_TC1</bpmndt:name>
          <bpmndt:path>
            <bpmndt:node>StartEvent_Participant_2</bpmndt:node>
            <bpmndt:node>Call_Participant_3</bpmndt:node>
            <bpmndt:node>EndEvent_Participant_2</bpmndt:node>
          </bpmndt:path>
        </bpmndt:testCase>
      </bpmndt:testCases>
    </bpmn:extensionElements>
    <bpmn:startEvent id="StartEvent_Participant_2" name="Start">
      <bpmn:outgoing>Flow_0i6rvwd</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:callActivity id="Call_Participant_3" name="Call Participant 3" calledElement="Participant_3_PID">
      <bpmn:incoming>Flow_0i6rvwd</bpmn:incoming>
      <bpmn:outgoing>Flow_0rqjk1w</bpmn:outgoing>
    </bpmn:callActivity>
    <bpmn:endEvent id="EndEvent_Participant_2" name="End">
      <bpmn:incoming>Flow_0rqjk1w</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:sequenceFlow id="Flow_0i6rvwd" sourceRef="StartEvent_Participant_2" targetRef="Call_Participant_3" />
    <bpmn:sequenceFlow id="Flow_0rqjk1w" sourceRef="Call_Participant_3" targetRef="EndEvent_Participant_2" />
  </bpmn:process>
  <bpmn:error id="Error_0397gvb" name="Error_ConnectionFailed" errorCode="ERR_CONN_FAILED" camunda:errorMessage="Connection to external server failed" />
  <bpmn:error id="Error_0esr758" name="Error_TargetEmailServerFailed" errorCode="ERR_TARGET_EMAIL_SERVER_FAILED" />
  <bpmn:error id="Error_0ywwuka" name="Error_ConnectionFailed" errorCode="ERR_CONN_FAILED" camunda:errorMessage="Connection to external server failed" />
  <bpmn:error id="Error_1bxwu7l" name="Error_VirusFound" errorCode="ERR_VIRUS_FOUND" camunda:errorMessage="Virus found in email" />
  <bpmn:error id="Error_1jap84z" name="Error_ContentNotParsable" errorCode="ERR_CONTENT_NOT_PARSABLE" camunda:errorMessage="Content of file is not a valid email " />
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Collaboration_045oqf3">
      <bpmndi:BPMNShape id="Participant_0pqyah4_di" bpmnElement="Participant_1" isHorizontal="true">
        <dc:Bounds x="160" y="80" width="710" height="290" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1t44l2n_di" bpmnElement="StartEvent_Participant_1">
        <dc:Bounds x="242" y="212" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="236" y="255" width="55" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_11ojvip_di" bpmnElement="EndEvent_Participant_1">
        <dc:Bounds x="672" y="212" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="680" y="255" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1mjnqu2_di" bpmnElement="Call_Participant_2">
        <dc:Bounds x="510" y="190" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_16gb0hl_di" bpmnElement="CollectElementsTask">
        <dc:Bounds x="340" y="190" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_19m8hkc_di" bpmnElement="Flow_19m8hkc">
        <di:waypoint x="278" y="230" />
        <di:waypoint x="340" y="230" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1hqadxe_di" bpmnElement="Flow_1hqadxe">
        <di:waypoint x="610" y="230" />
        <di:waypoint x="672" y="230" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0q1lqbh_di" bpmnElement="Flow_0q1lqbh">
        <di:waypoint x="440" y="230" />
        <di:waypoint x="510" y="230" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="Participant_1ielzr9_di" bpmnElement="Participant_3" isHorizontal="true">
        <dc:Bounds x="160" y="690" width="710" height="280" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0fiz1o2_di" bpmnElement="StartEvent_Participant_3">
        <dc:Bounds x="262" y="812" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="268" y="855" width="25" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1wgwi8f_di" bpmnElement="UserTaks_1">
        <dc:Bounds x="360" y="790" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_0hk1s1r_di" bpmnElement="EndEvent_Participant_3">
        <dc:Bounds x="512" y="812" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="520" y="855" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_08dpap9_di" bpmnElement="Flow_08dpap9">
        <di:waypoint x="298" y="830" />
        <di:waypoint x="360" y="830" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_00eq203_di" bpmnElement="Flow_00eq203">
        <di:waypoint x="460" y="830" />
        <di:waypoint x="512" y="830" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="Participant_1fl9tey_di" bpmnElement="Participant_2" isHorizontal="true">
        <dc:Bounds x="160" y="400" width="710" height="260" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1qgnr9m_di" bpmnElement="StartEvent_Participant_2">
        <dc:Bounds x="252" y="522" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="258" y="565" width="25" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_17xup7s_di" bpmnElement="Call_Participant_3">
        <dc:Bounds x="340" y="500" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1wm47ys_di" bpmnElement="EndEvent_Participant_2">
        <dc:Bounds x="502" y="522" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="510" y="565" width="20" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_0i6rvwd_di" bpmnElement="Flow_0i6rvwd">
        <di:waypoint x="288" y="540" />
        <di:waypoint x="340" y="540" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0rqjk1w_di" bpmnElement="Flow_0rqjk1w">
        <di:waypoint x="440" y="540" />
        <di:waypoint x="502" y="540" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```

</details>

<details>
<summary>ParticipantsTest.java</summary>

```java
package org.example.it;

import generated.participant_1_pid.TC_Participant1_TC1;
import generated.participant_2_pid.TC_Participant2_TC1;
import org.camunda.bpm.engine.test.assertions.bpmn.ProcessInstanceAssert;
import org.example.ExampleApp;
import org.example.it.utils.DelegateMock;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

@SpringBootTest(classes = {ExampleApp.class, BpmndtConfiguration.class}, webEnvironment = SpringBootTest.WebEnvironment.NONE)
public class ParticipantsTest {
    @RegisterExtension
    public TC_Participant1_TC1 tc1 = new TC_Participant1_TC1();

    @RegisterExtension
    public TC_Participant2_TC1 tc2 = new TC_Participant2_TC1();

    @Test
    public void testExecuteP1TC1() {
        tc1.handleCall_Participant_2().verifyParallel();
        tc1.handleCall_Participant_2().verifyLoopCount(3);

        tc1.createExecutor()
                // .withBusinessKey("bk")
                .withVariable("elements", List.of("a", "b", "c"))
                .withBean("collectElements", new DelegateMock())
                .verify(ProcessInstanceAssert::isEnded)
                .execute();
    }

    @Test
    public void testExecuteP2TC1() {
        tc2.handleCall_Participant_3().verify((piAssert, callActivityDefinition) -> {

        });


        tc2.createExecutor()
                // .withBusinessKey("bk")
                .verify(ProcessInstanceAssert::isEnded)
                .execute();
    }
}
```

</details>

Is there a known issue with the `bpmn-driven-testing` library in a Spring Boot environment? Do you recommend using Spring Boot for these tests, or would sticking with plain Java be preferable? We chose Spring Boot because our project uses it but for the tests we would also be fine using plain Java, I guess.

Thanks :) and best regards

**Comment by gclaussn** [comment]
hi @bennobuilder,

I think I know what the problem with your Spring Boot tests might be.

the [advanced-spring-boot](https://github.com/camunda-community-hub/bpmn-driven-testing/tree/master/integration-tests/advanced-spring-boot) integration test does not look like a Spring boot process application that I have usually under test - since it tests only the configration of the process engine, provided by the `camunda-bpm-spring-boot-starter`, when exposing the `BpmndtProcessEnginePlugin` as a Spring bean.

when running process tests (with or without bpmn-driven-testing), the `application.yaml` under src/test/resources must disable the job executor via:
```yaml
camunda:
  bpm:
    job-execution:
      enabled: false
```
(see https://docs.camunda.org/manual/latest/user-guide/spring-boot-integration/configuration/#camunda-engine-properties)

by default the job executor is enabled and runs in parallel in a separate thread. this makes the process tests not deterministic and leads to the exceptions you described (`No job found` or `Entity was updated by another transaction concurrently`). therefore it must be disabled. I will add this configuration to the Spring Boot integration test to make it complete.

please give me a quick feedback if this makes your Spring Boot based tests run successfully.

best regards

**Comment by bennobuilder** [comment]
Hi @gclaussn,

Thanks for your suggestion. I tested it in a demo project, and it works perfectly. However, for our main project, we decided to proceed without Spring Boot testing for now since it's easier to setup and doesn't seem to be necessary to test the BPMN, even though the project itself uses Spring Boot. Your library has made the overall testing process much smoother.

Thanks again, and cheers :)
